From 209a263ab2191f0b1d13f48a0a2fa67794357ad0 Mon Sep 17 00:00:00 2001
From: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>
Date: Fri, 30 Apr 2021 08:35:31 +0200
Subject: [PATCH] dvbconfigdialog: increase max download size for scandata

By the time Kaffeine was written, a 64KB max limit were
enough, but nowadays, the scandata file has already 104KB.

Increase the limit up to 1MB, in order to fix download
issues with valid files.

While here, also improve the download messages to show
how much data was downloaded and to report the reason
why a download was aborted.

BUG: 436371

Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>
---
 src/dvb/dvbconfigdialog.cpp | 28 +++++++++++++++++++++-------
 src/dvb/dvbconfigdialog.h   |  1 +
 2 files changed, 22 insertions(+), 7 deletions(-)

diff --git a/src/dvb/dvbconfigdialog.cpp b/src/dvb/dvbconfigdialog.cpp
index 7b45777..2abc606 100644
--- a/src/dvb/dvbconfigdialog.cpp
+++ b/src/dvb/dvbconfigdialog.cpp
@@ -614,6 +614,8 @@ DvbScanFileDownloadDialog::DvbScanFileDownloadDialog(DvbManager *manager_, QWidg
 	setLayout(mainLayout);
 	mainLayout->addWidget(mainWidget);
 
+	errorMsg = "Scan data update aborted.";
+
 	buttonBox = new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
 	buttonBox->button(QDialogButtonBox::Ok)->setEnabled(false);
 	connect(buttonBox, SIGNAL(accepted()), this, SLOT(accept()));
@@ -655,10 +657,20 @@ void DvbScanFileDownloadDialog::progressChanged(KJob *, unsigned long percent)
 
 void DvbScanFileDownloadDialog::dataArrived(KIO::Job *, const QByteArray &data)
 {
-	if ((scanData.size() + data.size()) <= (64 * 1024)) {
-		scanData.append(data);
-	} else {
-		job->kill(KJob::EmitResult);
+	scanData.append(data);
+
+	/*
+	 * Current size in April, 2021 is 104KB.
+	 * Add an upper size limit here (1 MB), just in case, in order to
+	 * prevent potential issues.
+	 */
+	int maxLimit = 1024 * 1024;
+
+	if (scanData.size() > maxLimit) {
+	    errorMsg = i18n("Scan file is becoming bigger than %1 KB.\n"
+			    "Aborting, as something might be wrong...\n",
+			    QString::number(maxLimit / 1024., 'f', 2));
+               job->kill(KJob::EmitResult);
 	}
 }
 
@@ -668,7 +680,7 @@ void DvbScanFileDownloadDialog::jobFinished()
 
 	if (job->error() != 0) {
 		if (job->error() == KJob::KilledJobError) {
-			label->setText(i18n("Scan data update failed."));
+			label->setText(errorMsg);
 		} else {
 			label->setText(job->errorString());
 		}
@@ -678,8 +690,10 @@ void DvbScanFileDownloadDialog::jobFinished()
 
 	if (manager->updateScanData(scanData)) {
 		buttonBox->button(QDialogButtonBox::Ok)->setEnabled(true);
-		label->setText(i18n("Scan data successfully updated. Changes take\n"
-				    "effect after you have closed the configuration dialog."));
+		label->setText(i18n("Scan data successfully updated. %1 KB received.\n"
+				    "Changes take effect after you have closed\n"
+				    "the configuration dialog.",
+			            QString::number(scanData.size() / 1024., 'f', 2)));
 	} else {
 		label->setText(i18n("Scan data update failed."));
 	}
diff --git a/src/dvb/dvbconfigdialog.h b/src/dvb/dvbconfigdialog.h
index 476afec..21994ea 100644
--- a/src/dvb/dvbconfigdialog.h
+++ b/src/dvb/dvbconfigdialog.h
@@ -138,6 +138,7 @@ private:
 	QByteArray scanData;
 	QVBoxLayout *mainLayout;
 	QDialogButtonBox *buttonBox;
+	QString errorMsg;
 };
 
 class DvbConfigPage : public QWidget
-- 
GitLab

